runtime/stdlib/cadence_v0.42_to_v1_contract_upgrade_validator.go

```diff
@@ -271,6 +283,24 @@ func (validator *CadenceV042ToV1ContractUpdateValidator) checkTypeUpgradability(
 			break
 		}
 		validator.currentRestrictedTypeUpgradeRestrictions = oldType.Types
+
+		// If the old restricted type is for AnyStruct/AnyResource,
+		// and if there are atleast one restriction, require them to drop the "restricted type".
+		// e.g-1: `AnyStruct{I} -> {I}`
+		// e.g-2: `AnyResource{I} -> {I}`
+		// See: https://github.com/onflow/cadence/issues/3112
+		if restrictedNominalType, isNominal := oldType.LegacyRestrictedType.(*ast.NominalType); isNominal {
+			switch restrictedNominalType.Identifier.Identifier {
+			case "AnyStruct", "AnyResource":
+				if len(oldType.Types) > 0 {
+					break typeSwitch
+				}
+			}
+		}
+
+		// Otherwise require them to drop the "restrictions".
+		// e.g-1: `T{I} -> T`
+		// e.g-2: `AnyStruct{} -> AnyStruct`
```

Comment:
How is it ensured this is only done in the case of an "owned" type / non-reference type (the parent is not a reference type)?

Reply:
yeah, this one does not; Need to update it to the match the changes we discussed for reference type.

Reply:
Actually, it is not needed for the contract update checker,  rather only needed for the static type migration, because entitlement migration is relying on the restriction set. Entitlement migration would anyway drop the restrictions after converting them to the corresponding entitlement.

So in the two cases, conversions would be:
- Reference type: `&T{I}` -> `auth(E) &T`
- Owned value: `T{I}` -> `T`

Both in which the restrictions are dropped.

Reply:
Then there is a question, what should `&AnyStruct{I}` be converted to? is it `auth(E) &AnyStruct` or `auth(E) &{I}`

The update checker currently forces it to be `auth(E) &{I}`. Do you know how the storage migration handles it?

Reply:
> Actually, it is not needed for the contract update checker, rather only needed for the static type migration, because entitlement migration is relying on the restriction set. 

Ah, right! The rules for rewriting intersection types I had written up in the Notion doc were specifically for the static type migration, considering that the entitlements migration would run after, requiring enough information to be preserved. 

But really, in the end the state migration (static type + entitlements migration) and the contract update checker should have rules for the same outcome at the end.

Reply:
> Then there is a question, what should `&AnyStruct{I}` be converted to? is it `auth(E) &AnyStruct` or `auth(E) &{I}`. The update checker currently forces it to be `auth(E) &{I}`.

I'll see what the entitlements migration currently produces. I guess `AnyStruct{I} -> {I}` should apply, i.e. `&AnyStruct{I}` should be rewritten to `auth(E) &{I}` (instead of just `auth(E) &AnyStruct`), but IDK if that's currently the case

Reply:
I'll fix this in #3118